src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = []; // empty array that holds history of changes
+  let currentState = { ...state }; // deep copy of the initial state object
+
+  actions.forEach(action => {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = {
+          ...currentState, ...action.extraData, // copy the current state
+          // and adds extra action
+        };
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach(key => delete currentState[key]);
+        break;
+      case 'clear':
+        currentState = {};
+        break;
+      default:
+        throw new Error();
```

Comment:
Usually it's a bad idea to throw without any context, because this means that consumer of this function (You in the different part of the code, or someone else in the team) has to know that this function is not safe and has to check if it will not throw.
I would change it to `console.error('Your message')`
You can disable linting for this line only.